pkg/component/runtime/service_command.go

```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
+	go func() {
+		retryAttempt := 0
+
+		// Here we indefinitely retry the executeCommand call, as long as it
+		// returns a non-nil error. We will block here until executeCommand
+		// returns a nil error, indicating that the command being executed has
+		// successfully completed execution.
+		//nolint: errcheck // No point checking the error inside the goroutine.
+		backoff.RetryNotify(
+			func() error {
+				err := executeCommand(cmdCtx, log, binaryPath, args, env, timeout)
+				cmdDone <- struct{}{}
+				return err
+			},
+			backoffCtx,
+			func(err error, retryAfter time.Duration) {
+				retryAttempt++
+				log.Warnf(
+					"service command execution failed with error [%s], retrying (will be retry [%d]) after [%s]",
+					err.Error(),
+					retryAttempt,
+					retryAfter,
+				)
+				<-cmdDone
+			},
+		)
+
+		cr.untrack(cmdKey)
+	}()
+}
+
+func (cr *cmdRetrier) Stop(cmdKey uint64, log *logger.Logger) {
+	cr.mu.RLock()
+	defer cr.mu.RUnlock()
+	info, exists := cr.cmds[cmdKey]
+	if !exists {
+		log.Debugf("no retries for command key [%d] are pending; nothing to do", cmdKey)
+		return
+	}
+
+	// Cancel the previous retries
+	info.retryCancelFn()
+
+	// Cancel the previous command
+	info.cmdCancelFn()
+
+	// Ensure that the previous command actually stopped running
+	<-info.cmdDone
+
+	// Stop tracking
+	delete(cr.cmds, cmdKey)
+	log.Debugf("retries and command process for command key [%d] stopped", cmdKey)
+}
+
+func (cr *cmdRetrier) track(cmdKey uint64, cmdCancelFn context.CancelFunc, retryCanceFn context.CancelFunc, cmdDone <-chan struct{}) {
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	// Initialize map if needed
+	if cr.cmds == nil {
+		cr.cmds = map[uint64]cmdRetryInfo{}
+	}
+
+	cr.cmds[cmdKey] = cmdRetryInfo{
+		retryCancelFn: retryCanceFn,
+		cmdCancelFn:   cmdCancelFn,
+		cmdDone:       cmdDone,
+	}
+}
+
+func (cr *cmdRetrier) untrack(cmdKey uint64) {
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	delete(cr.cmds, cmdKey)
+}
+
+// cmdKey returns a unique, deterministic integer for the combination of the given
+// binaryPath, args, and env. This integer can be used to determine if the same command
+// is being executed again or not.
+func (cr *cmdRetrier) cmdKey(binaryPath string, args []string, env []string) uint64 {
+	var sb strings.Builder
+
+	sb.WriteString(binaryPath)
+	sb.WriteString("|")
+
+	var posArgs, flagArgs []string
+	for i := 0; i < len(args); i++ {
+		arg := args[i]
+		if strings.HasPrefix(arg, "-") {
+			flagArgs = append(flagArgs, arg+" "+args[i+1])
+			i++
+		} else {
+			posArgs = append(posArgs, arg)
+		}
+	}
+
+	sort.Strings(posArgs)
+	sort.Strings(flagArgs)
+
+	for _, arg := range posArgs {
+		sb.WriteString(arg)
+		sb.WriteString("|")
+	}
+
+	for _, arg := range flagArgs {
+		sb.WriteString(arg)
+		sb.WriteString("|")
+	}
+
+	sort.Strings(env)
+
+	for _, kv := range env {
+		sb.WriteString(kv)
+		sb.WriteString("|")
+	}
+
+	digest := fnv.New64()
+	digest.Write([]byte(sb.String()))
+
+	return digest.Sum64()
```

Comment:
> looking at the fnv1 and fnv1a there's a high probability of collisions, how do we want to handle that?

Where are you seeing this? 
 I looked at https://github.com/Cyan4973/xxHash/wiki/Collision-ratio-comparison and fnv looked decent to me.  I've used `xxHash` in the past but I didn't want to pull in another dependency just for the usage here.  Do you have any recommendation for an alternate hashing algorithm that has a lower probability of collisions than fnv?

Reply:
https://theses.liacs.nl/pdf/2014-2015NickvandenBosch.pdf

Reply:
i'm just thinknig how many concurrent intalls/uninstalls can one have. under any circumestantces there should not be different versions of service being installed so spec args are the same for service type always. maybe we dont even need this type of hashing. if new install is starting it will be install of the exactly same service as the one already running.

Reply:
The issue is that at this level of the code path, we don't know which service operation (`install`, `uninstall`) is being run.  So it's possible that the command for the `install` operation is already running when the command for `uninstall` is called.  In this scenario, I think we'd want the running `install` operation command to finish, that is, we would not want to stop it.  So I think, to be safe, it's better to have the hashing or some other way to know that the command being run is exactly the same as a command already running or not.

Thanks for sharing the paper.  Looking at the table in there, I think SHA256 with 64-bit output should suffice.

Reply:
i mean yes that's true, but it costs us a few cycles. but this not being a hot path i think we should be fine